### PR TITLE
[Feature-12040][api][ui] Add authorization management of read and write permissions for project center

### PR DESCRIPTION
**Pull request by GavinGYM**
## Purpose of the pull request

At present, DolphinScheduler's control of permissions is relatively simple, and it is hoped to increase the control of operation permissions to separate read and write permissions when authorizing projects.

## Brief change log

- Distinguish read and write permissions for the entire content of a single project, and grant read-only or read-write permissions to a single user for a single project
- Add read-only and read-write authorization operations to the front-end page when authorization is performed

## Verify this pull request

This change added tests and can be verified as follows:

- Added unit tests for grantProjectWithReadPerm, revokeProjectById and grantResourceWithPermLevel API in UsersServiceTest.java.
- Added a unit test for hasProjectAndWritePerm API in ProjectServiceTest.java.

### 1. Admin can now grant read-only or read & write permissions to a single user for a single project
Go to Security→User Manage→Operation→Authorize→Project, and you will see the new Project Authorize window, where admin user can grant different permissions for a project to a user, by simply selecting projects and clicking different authorization buttons. And I also implemented paging and query functions for ease of use. 
Untitled
Untitled (1)
### 2. Read and write permissions for the entire content of a single project are distinguished
If a user has only the read permission but not the write permission for an project, and the user is trying to do something like delete or update the project, an error message is displayed indicating that the user has no write permission and cannot complete the operation.
Untitled (2)

### CODE REVIEW

**Comment by GavinGYM** [comment]
## Screenshots of newly added features
### 1. Admin can now grant read-only or read & write permissions to a single user for a single project
Go to Security→User Manage→Operation→Authorize→Project, and you will see the new Project Authorize window, where admin user can grant different permissions for a project to a user, by simply selecting projects and clicking different authorization buttons. And I also implemented paging and query functions for ease of use. 
Untitled
Untitled (1)
### 2. Read and write permissions for the entire content of a single project are distinguished
If a user has only the read permission but not the write permission for an project, and the user is trying to do something like delete or update the project, an error message is displayed indicating that the user has no write permission and cannot complete the operation.
Untitled (2)
### 3. Admin can now grant read-only or read & write permissions to a single user for a single item in resource center.
Go to Security→User Manage→Operation→Authorize→Resource, and you will see the new Resource Authorize window, where admin user can grant different permissions for Resources Center items (including File Resource and UDF Resource) to a user. I kept the original Tree Select, and added a button to switch between different levels of permissions granted.
Untitled (3)
Untitled (4)
### 4. Read and write permissions for all content in the resource center are distinguished.
If a user has only the read permission but not the write permission for a content in the resource center, and the user is trying to do something like delete or update the resource, an error message is displayed indicating that the user has no write permission and cannot complete the operation.
Untitled (5)

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java [hunk 1]
```diff
@@ -413,12 +497,56 @@ public Result update(User loginUser, Long projectCode, String projectName, Strin
...
+     *
+     * @param loginUser login user
+     * @return project list
+     */
+    @Override
+    public Map<String, Object> queryProjectWithAuthorizedLevel(User loginUser, Integer userId) {
+        Map<String, Object> result = new HashMap<>();
+
+        Set<Integer> projectIds = resourcePermissionCheckService.userOwnedResourceIdsAcquisition(AuthorizationType.PROJECTS, loginUser.getId(), logger);
+        //        if (projectIds.isEmpty()) {
```
Comment (zixi0825) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java]: Please remove the useless code

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -254,6 +264,33 @@ public boolean hasProjectAndPerm(User loginUser, Project project, Map<String, Ob
...
+            if (loginUser.getUserType() == UserType.ADMIN_USER) {
+                return true;
+            }
+            // case 2: user is project owner
+            if (project.getUserId() == loginUser.getId()) {
+                return true;
+            }
+            // case 3: check user permission level
+            ProjectUser projectUser = projectUserMapper.queryProjectRelation(project.getId(), loginUser.getId());
+            // if (!canOperatorPermissions(loginUser, new Object[]{project.getId()},AuthorizationType.PROJECTS,perm)){
```
Comment (zixi0825) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java]: Please remove the useless code

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java [hunk 3]
```diff
@@ -680,6 +788,149 @@ public Map<String, Object> revokeProject(User loginUser, int userId, long projec
...
+    /**
+     * grant resource with permission level
+     *
+     * @param loginUser           login user
+     * @param userId              user id
+     * @param readPermResourceIds resource id array with read permission
+     * @param allPermResourceIds  resource id array with all permission
+     * @return grant result code
+     */
+    @Override
```
Comment (zixi0825) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java]: The method body is too long, please refactor

File: dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts [hunk 4]
```diff
@@ -217,10 +316,10 @@ export function useAuthorize() {
     if (state.saving) return false
     state.saving = true
     if (type === 'authorize_project') {
-      await grantProject({
-        userId,
-        projectIds: state.authorizedProjects.join(',')
-      })
+      // await grantProject({
```
Comment (zixi0825) [inline dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts]: Please remove the useless code

**Comment by zixi0825** [comment]
@Amy0104 cc

**Review by SbloodyS — DISMISSED** [review]
The backend part LGTM. Would you please add this new feature to the documents?

**Comment by GavinGYM** [comment]
> The backend part LGTM. Would you please add this new feature to the documents?

I'm glad to. I'll figure out where to put it in the document later.

File: dolphinscheduler-ui/src/service/modules/projects/index.ts [hunk 5]
```diff
@@ -26,6 +26,14 @@ export function queryProjectListPaging(params: ListReq): any {
   })
 }
 
+export function queryProjectWithAuthorizedLevelListPaging(params: any): any {
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/service/modules/projects/index.ts]: It's better to define the type for the params.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx [hunk 6]
```diff
@@ -43,14 +50,14 @@ const props = {
     default: 'auth_project'
   }
 }
-
+let projectIds: string
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: why define the projectIds here not inside the setup?

File: dolphinscheduler-ui/src/service/modules/projects/index.ts (hunk 5 above)
Comment (GavinGYM) [reply dolphinscheduler-ui/src/service/modules/projects/index.ts]: Suggestions received and have been revised as suggested.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx (hunk 6 above)
Comment (GavinGYM) [reply dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: In fact, I'm not particularly familiar with the front-end, so this part of the front-end code implementation I asked a friend to help. Regarding this declaration, I noticed that putting it in setup may causes errors at lines 116,119, and 122. If you have a better implementation, please do not hesitate to comment.

Comment (Amy0104) [reply dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: Pass it to the render function by the return of the setup, then you can use it in the render. But I think it's better to call the functions(such as `revokeProjectByIdRequest `) in the setup function.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx [hunk 7]
```diff
@@ -88,14 +109,66 @@ export const AuthorizeModal = defineComponent({
...
             filterable
             v-model={[this.authorizedProjects, 'value']}
             class={styles.transfer}
           />
+        )} */}
+        {type === 'authorize_project' && (
+
+          <NSpace vertical>
+            <NSpace>
+              <NButton size='small' type='primary' onClick={() => this.revokeProjectByIdRequest(userId, projectIds)}>
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: BTW, if you are using an arrow function, and you don't want to use curly brackets. It's better to add `void` after the arrow that means the arrow function return nothing.

**Comment by zixi0825** [comment]
LGTM

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx (hunk 6 above)
Comment (GavinGYM) [reply dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx:53]: I tried to rewrite this part according to your suggestion. However, maybe due to my poor front-end skill, after changing this, functions such as revokeProjectByIdRequest will get projectIds as undefined value, I tried but couldn't fix this problem.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx (hunk 7 above)
Comment (GavinGYM) [reply dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: Suggestions received and have been revised as suggested.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx (hunk 6 above)
Comment (GavinGYM) [reply dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: @Amy0104 Suggestions received and have been revised as suggested in the latest commit. I define the _projectIds_ inside the setup, and call the functions(such as _revokeProjectByIdRequest_) in the _setup_ function. Please review again.

File: dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts [hunk 8]
```diff
@@ -56,29 +60,77 @@ export function useAuthorize() {
...
+    pagination: {
+      pageSize: 5,
+      page: 1,
+      totalPage: 0
+    },
+    searchVal: '',
+    userId: 0
   })
 
+  const onOperationClick = ( ) => { }
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts]: Is this function unused? If unused, plz remove it.

File: dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts [hunk 9]
```diff
@@ -217,10 +279,7 @@ export function useAuthorize() {
     if (state.saving) return false
     state.saving = true
     if (type === 'authorize_project') {
-      await grantProject({
-        userId,
-        projectIds: state.authorizedProjects.join(',')
-      })
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts:223]: This if statement seems to be useless.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx [hunk 10]
```diff
@@ -59,6 +65,21 @@ export const AuthorizeModal = defineComponent({
       if (result) onCancel()
     }
 
+    let projectIds!: string
+    const onRevokeProject = async () => {
+      await revokeProjectByIdRequest(props.userId, projectIds)
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: It seems to not need a async function here.

File: dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx [hunk 11]
```diff
@@ -149,6 +228,30 @@ export const AuthorizeModal = defineComponent({
...
+              filterable
+              multiple
+              cascade
+              checkable
+              checkStrategy='child'
+              key-field='id'
+              label-field='fullName'
+              options={this.udfResources}
+              v-model:value={this.authorizedUdfResourcesWithReadPerm}
+            />
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/authorize-modal.tsx]: It is better to control the availability of permission by data.

File: dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts [hunk 12]
```diff
@@ -265,10 +324,42 @@ export function useAuthorize() {
...
+          const arr = []
+          arr[0] = v1
+          if (getParent(arr, v).length > 0) {
+            fullPathUdfIdWithReadPerm = getParent(arr, v).map((v2: any) => {
+              return v2.id
+            })
+            pathUdfIdWithReadPerm.push(fullPathUdfIdWithReadPerm.join('-'))
+          }
+        })
+      })
```
Comment (Amy0104) [inline dolphinscheduler-ui/src/views/security/user-manage/components/use-authorize.ts]: This block is similar to the previous one. It's better to do this by a function.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java [hunk 13]
```diff
@@ -533,6 +533,97 @@ public Map<String, Object> deleteUserById(User loginUser, int id) throws IOExcep
...
+        User tempUser = userMapper.selectById(userId);
+        if (tempUser == null) {
+            putMsg(result, Status.USER_NOT_EXIST, userId);
+            return result;
+        }
+
+        if (check(result, StringUtils.isEmpty(projectIds), Status.SUCCESS)) {
+            return result;
+        }
+        Arrays.stream(projectIds.split(",")).distinct().forEach(projectId -> {
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java]:
```suggestion
        Arrays.stream(projectIds.split(Constants.COMMA)).distinct().forEach(projectId -> {
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java [hunk 14]
```diff
@@ -685,6 +778,139 @@ public Map<String, Object> revokeProject(User loginUser, int userId, long projec
...
+    /**
+     * Change the resource ids from String to Set
+     *
+     * @param ResourceIds resource id array that need authorize
+     * @return Integer Set of resource ids
+     */
+    private Set<Integer> getNeedAuthorizeResIds(String ResourceIds) {
+        Set<Integer> needAuthorizeResIds = new HashSet<>();
+        if (StringUtils.isNotBlank(ResourceIds)) {
+            String[] resourceFullIdArr = ResourceIds.split(",");
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java]:
```suggestion
            String[] resourceFullIdArr = ResourceIds.split(Constants.COMMA);
```

**Review by Amy0104 — APPROVED** [review]
The front end part LGTM.

**Comment by caishunfeng** [comment]
@GavinGYM this pr miss updating docs, please submit another doc pr for it, thanks.  cc @zixi0825

**Comment by GavinGYM** [comment]
> @GavinGYM this pr miss updating docs, please submit another doc pr for it, thanks. cc @zixi0825

@caishunfeng Gotcha, will complete the document in a few days.

**Pull request merged**